Add device interface specification (#215)
harp-tech/protocol: DeviceInterface.md

## Requirements Language

The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT", "SHOULD", "SHOULD NOT", "RECOMMENDED", "MAY", and "OPTIONAL" in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt).

## Relationship to Other Specifications

- [**Binary Protocol**](Protocol.md) — defines the wire format for all Harp messages (message type, payload type encoding, checksum, etc.).
- [**Device Registers and Operation**](Device.md) — defines the behavioral contract every Harp device must follow (operation modes, request-reply patterns, core registers). The device interface file describes *which* registers a specific device exposes, not *how* the protocol operates. Additionally, the device interface further specifies application registers and their payloads, which are outside the scope of the core device specification.

## Document Structure Overview

> A device interface file is a YAML document whose schema is defined by `./schema/device.json`. The root level of the document contains metadata fields describing the device and a `registers` map that declares all application registers. Optional `bitMasks` and `groupMasks` sections define named masks that can be referenced by registers.

Reviewer: So the first line duplicates/contradicts the "Schema Reference" at the end of the file, which states that it's defined by a set of 3 files (`device.json`, `core.json`, `registers.json`). I think it might help to bring the "Schema Reference" section with the table up here and delete this line, as I also find that section is a little out of place at the end of the article. 

I also realize the term `device.yml` is never used anywhere, while `core.yml` is mentioned. I feel like it can be brought in here.

```suggestion
The root level of the device interface file (`device.yml`) contains metadata fields describing the device and a `registers` map that declares all application registers. Optional `bitMasks` and `groupMasks` sections define named masks that can be referenced by registers.
```

The file MUST declare all four metadata fields (`device`, `whoAmI`, `firmwareVersion`, `hardwareTargets`) and the `registers` map. The `bitMasks` and `groupMasks` sections are OPTIONAL and SHOULD be included when registers reference named masks.

> A YAML language-server schema directive SHOULD be placed at the top of the file to enable editor validation and autocompletion:

Reviewer: ```suggestion
A YAML version and language-server schema directive SHOULD be placed at the top of the file to enable editor validation and autocompletion:
```

Include the YAML version directive here?

```yaml

> # yaml-language-server: $schema=https://harp-tech.org/draft-02/schema/device.json

Reviewer: ```suggestion
%YAML 1.1
---
# yaml-language-server: $schema=https://harp-tech.org/draft-03/schema/device.json
```

Add yaml version directive and update the schema directive link which is now on draft-03

| `device`          | string  |   Yes    | The human-readable name of the device (e.g. `Behavior`).                                                                                                                                 |
| `whoAmI`          | integer |   Yes    | The unique identity class of the device, as registered in the [harp-tech/whoami](https://github.com/harp-tech/whoami) repository. Corresponds to core register `R_WHO_AM_I` (address 0). |
| `firmwareVersion` | string  |   Yes    | The firmware version in `"<major>.<minor>"` format (e.g. `"3.4"`).                                                                                                                       |
| `hardwareTargets` | string  |   Yes    | The hardware version targeted by this firmware, in `"<major>.<minor>"` format (e.g. `"2.0"`).                                                                                            |

**Example:**

```yaml
device: Behavior
whoAmI: 1216
firmwareVersion: "3.4"
hardwareTargets: "2.0"
```

## Core Registers

> Every Harp device also exposes a set of core registers (addresses 0–31) that are common to all devices. These registers are defined in a separate file (`core.yml`) and follow the same YAML schema described in this document. Core registers handle device identification, timestamping, operation mode control, and other shared functionality. See [Device Registers and Operation](Device.md#core-registers) for the full specification of core register behavior.

Reviewer: ```suggestion
Every Harp device also exposes a set of core registers (addresses 0–31) defined in a separate [`core.yml`](./schema/core.yml) file that follows the same YAML schema described in this document. Core registers handle device identification, timestamping, operation mode control, and other shared functionality. As these registers are common to all devices, `core.yml` does not need to be included alongside `device.yml`. See [Device Registers and Operation](Device.md#core-registers) for the full specification of core register behavior.
```

I feel like one thing that is not obvious to people is whether or not they should include core.yml in the device repository, so I suggested edits here, and added link to schema/core.yml.

## Registers

The `registers` section is a map where each key is the register name (PascalCase by convention) and the value is a register definition object.

### Register Properties

| Property        | Type               | Required | Default  | Description                                                                                                                     |
| :-------------- | :----------------- | :------: | :------: | :------------------------------------------------------------------------------------------------------------------------------ |
| `address`       | integer (0–255)    |   Yes    |    —     | The unique 8-bit address of the register. Application registers MUST use addresses ≥ 32.                                        |
| `type`          | string             |   Yes    |    —     | The payload type. One of: `U8`, `S8`, `U16`, `S16`, `U32`, `S32`, `U64`, `S64`, `Float`.                                        |
| `length`        | integer (≥ 1)      |    No    |    1     | The number of elements in the register payload. Values greater than 1 indicate an array payload.                                |

> | `access`        | string or string[] |   Yes    |    —     | The access mode(s) of the register. A single value or an array of 1–3 unique values from: `Read`, `Write`, `Event`.            |

Reviewer: ```suggestion
| `access`        | string or string[] |   Yes    |    —     | The access mode(s) of the register. A single value or an array of 1–3 unique values from: `Read`, `Write`, `Event`.            |
| `length`        | integer (≥ 1)      |    No    |    1     | The number of elements in the register payload. Values greater than 1 indicate an array payload.                                |
```

Is it possible to move the `access` property up? It feels more intuitive to list the required registers first.

### Access Modes

Every register MUST declare an `access` field, which determines which Harp message types are valid for it:

| Access  | Behavior                                                                                                                          |
| :------ | :-------------------------------------------------------------------------------------------------------------------------------- |
| `Read`  | The register supports `Read` requests from the Controller. The device replies with the current register value.                    |
| `Write` | The register supports `Write` requests from the Controller. The device updates the register and replies with the resulting value. |
| `Event` | The device MAY send unsolicited `Event` messages from this register when a relevant trigger occurs.                               |

A register that specifies only `Event` as its access mode is implicitly read-only from the Controller's perspective — the Controller cannot write to it, and the device pushes data when events occur.

> When multiple access modes are needed, they SHOULD be expressed as a YAML sequence (1–3 unique values):

Reviewer: ```suggestion
When multiple access modes are needed, they MUST be expressed as a YAML sequence (1–3 unique values):
```

Is there another way of expressing multiple access modes? If not, should it be a MUST?

```yaml
access: [Write, Event]
```

### Value Constraints

The `minValue` and `maxValue` fields specify the allowable range for the register payload. The device MAY reject writes outside this range (see [Device.md — Error Handling](Device.md#error-handling)).

**Example:**

```yaml
PwmDutyCycleDO0:
  address: 64
  type: U8
  access: Write
  minValue: 1
  maxValue: 99
  description: Specifies the duty cycle of the PWM at DO0.
```

### Mask Type Reference

When a register's payload represents a set of flags or an enumerated value, the `maskType` field SHOULD reference a named mask defined in the [`bitMasks`](#bit-masks) or [`groupMasks`](#group-masks) section. This enables code generators to produce strongly-typed enumerations.

**Example:**

```yaml
DigitalInputState:
  address: 32
  access: Event
  type: U8
  maskType: DigitalInputs
  description: Reflects the state of DI digital lines of each Port
```

> ### Simple Register Examples
> 
> A minimal read-only register:
> 
> ```yaml
> InfoRegister:
>   address: 33
>   type: U8
>   access: Read
>   description: A simple read-only register that returns a status code.
> ```
> 
> A write-only register with value constraints:
> 
> ```yaml
> PulseDurationDOPort0:
>   address: 46
>   type: U16
>   access: Write
>   minValue: 1
>   description: Specifies the duration of the output pulse in milliseconds.
> ```

Reviewer: ```suggestion

```

Not sure if these simple examples are needed, as there are many other register examples, and the write-only example is very similar to the example in the "Value Constraints" section.  Maybe if you really want to keep the read-only register example it can be shifted to the "Access Modes" section.

| `length`        | integer |    No    | The number of elements this member occupies (default: 1).                    |
| `mask`          | integer |    No    | The bitmask used to read and write this member within a single payload word. |
| `description`   | string  |    No    | A summary description of the payload member.                                 |
| `minValue`      | number  |    No    | The minimum allowable value for this member.                                 |
| `maxValue`      | number  |    No    | The maximum allowable value for this member.                                 |
| `defaultValue`  | number  |    No    | The default value for this member.                                           |
| `maskType`      | string  |    No    | The name of a mask used to interpret this member's value.                    |
| `interfaceType` | string  |    No    | The type name used for this member in high-level interfaces.                 |

A member is located within the payload using two **independent** coordinates, either or both of which MAY be present:

- `offset` (with optional `length`) selects *which element(s)* of the payload array the member occupies.

> - `mask` selects *which bits* within a payload word the member occupies.

Reviewer: So one thing that really tripped me up is I am not sure when one should define the bitmask in the `payloadSpec` and when one should define it in the "bitMask" section of the `device.yml`. Are these reasons correct?

Define bitmask in payloadSpec:
- When combining offset and bitmask to locate a member?
- When separating heterogenous type members with different-sized bitfields?

Define bitmask in bitMasks section:
- When defining the same type members, with equal-size bitfields?

I also think I may have the wrong mental model and might be completely off about the distinction between the two 😆 A note about the differences between the two might help other confused readers like me.

```

`interfaceType` MAY also be applied at the **register level** (with no `payloadSpec`), in which case the whole payload is reinterpreted:

```yaml
CustomPayload:
  address: 36
  type: U32
  length: 3
  access: Read
  interfaceType: HarpVersion
```

> [!NOTE]
> The set of valid `interfaceType` names is intentionally open-ended (the schema treats it as a free string). Values seen in practice include primitive type names such as `bool`, `byte`, `int`, `uint`, `ushort`, `float`, and `string`, as well as custom type names such as `HarpVersion`. The meaning of each name is defined by the consuming tooling, not by this schema.

### Interface Types

> [!WARNING]
> The canonical list of supported (standardized) `interfaceType` values listed below. We MAY add additional types in the future if widespread use cases emerge.

| Interface Type | Notes                                              |
| :------------- | :------------------------------------------------- |
| `bool`         | Boolean flag, typically over a single masked bit.  |
| `byte`         | 8-bit unsigned integer.                            |
| `int`          | Signed integer.                                    |

> | `uint`         | Unsigned integer.                                  |

Reviewer: ```suggestion
| `int`          | 32-bit signed integer.                             |
| `uint`         | 32-bit unsigned integer.                           |
```

| `ushort`       | 16-bit unsigned integer.                           |
| `float`        | Floating-point value.                              |
| `string`       | Text, typically over a span of byte elements. When using interfaceType, the `length` parameter MUST be defined     |

> | `HarpVersion`  | Custom type composed from multiple elements that describes a SemVer-like version. When using this `interfaceType`, the `length` MUST be equal to 3       |

Reviewer: ```suggestion
| `HarpVersion`  | Custom type composed from multiple elements that describes a [semantic version](https://semver.org/). When using this `interfaceType`, the `length` MUST be equal to 3       |
```

Changed text and added link, similar to what was done in Device.MD

      interfaceType: ushort
StartPulseTrain:
  address: 101
  type: U16
  length: 2
  access: Write
  payloadSpec:
    <<: *startPulse        # inherits DigitalOutput and PulseWidth
    Frequency:             # and adds further members
      offset: 1
      mask: 0xFF00
      interfaceType: byte
```

### Anchor Scoping

Anchors MAY reference values as well as objects. For example, a common payload type can be shared across registers using a value anchor:

```yaml
Rgb0:
  address: 71
  type: &rgbType U8
  length: 3
  access: &rgbAccess Write
  description: Specifies the state of the RGB0 LED channels.
Rgb1:
  address: 72
  type: *rgbType
  length: 3
  access: *rgbAccess
  description: Specifies the state of the RGB1 LED channels.
```

### Private registers

Registers that are not meant to be exposed in high-level interfaces (e.g. code-generated classes) can be marked with `visibility: private`. This is commonly used for reserved registers that fill gaps in the address space to maintain contiguity.

```yaml
Reserved0: &reserved
  address: 33
  type: U8
  access: Read
  description: Reserved for future use
  visibility: private
Reserved1:
  <<: *reserved
  address: 84

> ```

Reviewer: Hmm, I feel like this section might be better placed in the walkthrough of the properties in the "Registers"  section. The example will have to be changed to get rid of the yaml anchors and aliases.

## Schema Reference

The device interface YAML format is formally defined by a set of JSON Schema files:

| Schema           | Description                                                                                                 |
| :--------------- | :---------------------------------------------------------------------------------------------------------- |
| `device.json`    | Top-level schema for device interface files. Requires device metadata and references `core.json`.           |
| `core.json`      | Schema for core device properties. Requires `protocolVersion` and references `registers.json`.              |
| `registers.json` | Defines the structure of registers, bit masks, group masks, payload members, and all associated properties. |

> Editors supporting the [YAML Language Server](https://github.com/redhat-developer/yaml-language-server) can validate device interface files by adding the schema directive shown in the [Document Structure Overview](#document-structure-overview).

Reviewer: As mentioned there is some overlap with this and the document overview section so might be better to move this up there.
